### PR TITLE
TapDS API

### PR DESCRIPTION
**Pull request by yuval-k**
API part of envoyproxy/envoy#7194

This introduces the API for the changes coming for the tap filter. 
The main features are adding a grpc server sink, and a configuration via xDS.

Risk Level: Low (just API - no implementation)
Testing: N/A
Docs Changes: docs are there but hidden.
Release Notes: N\A
Part of envoyproxy/envoy#6110

### CODE REVIEW

File: api/envoy/config/common/tap/v2alpha/common.proto [hunk 1]
```diff
@@ -14,6 +15,16 @@ option java_package = "io.envoyproxy.envoy.config.common.tap.v2alpha";
 
 // Common configuration for all tap extensions.
 message CommonExtensionConfig {
+
+  // [#not-implemented-hide:]
+  message TapDSConfig {
+    // Configuration for the source of EDS updates for this Cluster.
```
Comment (mattklein123) [inline api/envoy/config/common/tap/v2alpha/common.proto]: s/EDS/TapDS

File: api/envoy/config/common/tap/v2alpha/common.proto [hunk 2]
```diff
@@ -14,6 +15,16 @@ option java_package = "io.envoyproxy.envoy.config.common.tap.v2alpha";
...
+    envoy.api.v2.core.ConfigSource config_source = 1;
+
+    // Tap config to request from XDS server.
+    string name = 2;
```
Comment (mattklein123) [inline api/envoy/config/common/tap/v2alpha/common.proto]: non-empty string validation

File: api/envoy/data/tap/v2alpha/wrapper.proto [hunk 3]
```diff
@@ -13,9 +15,38 @@ option java_package = "io.envoyproxy.envoy.data.tap.v2alpha";
 
 // [#protodoc-title: Tap data wrappers]
 
+// [#not-implemented-hide:] Information about the endpoints involved in the tap.
+message Endpoints {
```
Comment (mattklein123) [inline api/envoy/data/tap/v2alpha/wrapper.proto]: Can we split this out into a different change? Meaning, I think this is orthogonal to TapDS and applies to all tap types. This will allow us to get streaming in first and then we can add additional data?

File: api/envoy/service/tap/v2alpha/common.proto [hunk 4]
```diff
@@ -24,6 +27,15 @@ message TapConfig {
   // a tap will occur and the data will be written to the configured output.
   OutputConfig output_config = 2 [(validate.rules).message.required = true];
 
+  // [#not-implemented-hide:] Specify if TAP is enabled. The % of requests\connections for which the
+  // filter is enabled.
```
Comment (mattklein123) [inline api/envoy/service/tap/v2alpha/common.proto]: tapping is currently supported on both filters and transport sockets and this config is shared. Can you generalize the name to not include "filter"?

File: api/envoy/service/tap/v2alpha/common.proto [hunk 5]
```diff
@@ -64,9 +76,17 @@ message MatchPredicate {
 
     // HTTP response trailers match configuration.
     HttpHeadersMatch http_response_trailers_match = 8;
+
+    // The destination cluster match configuration.
+    ClusterMatch destination_cluster_match = 9;
```
Comment (mattklein123) [inline api/envoy/service/tap/v2alpha/common.proto]: split out into a different change.

File: api/envoy/service/tap/v2alpha/tap.proto [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+// [#protodoc-title: Tap Sink Service]
+
+// [#not-implemented-hide:] Stream message for the Tap API. Envoy will open a stream to the server
+// and stream taps without ever expecting a response.
+message StreamTapsRequest {
+  message Identifier {
+    // The node sending taps over the stream.
+    envoy.api.v2.core.Node node = 1 [(validate.rules).message.required = true];
+    // The opaque identifier that was set in the output config.
```
Comment (mattklein123) [inline api/envoy/service/tap/v2alpha/tap.proto]: cross link

File: bazel/README.md [hunk 7]
```diff
@@ -40,7 +40,7 @@ for how to update or override dependencies.
 
     On Fedora (maybe also other red hat distros), run the following:
     ```
-    dnf install cmake libtool libstdc++ ninja-build lld patch
+    dnf install cmake libtool libstdc++ ninja-build lld patch aspell-en
```
Comment (mattklein123) [inline bazel/README.md]: unrelated change, please do a different PR

File: api/envoy/service/tap/v2alpha/tapds.proto [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+      body: "*"
+    };
+  }
+}
+
+// [#not-implemented-hide:] A tap resource is essentially a tap configuration with a name
+// The filter TapDS config references this name.
+message TapResource {
+  // The name of the tap configuration.
+  string name = 1;
```
Comment (mattklein123) [inline api/envoy/service/tap/v2alpha/tapds.proto]: non-empty string validation

File: api/envoy/service/tap/v2alpha/tap.proto [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    string tap_id = 2;
+  }
+
+  // Identifier data effectively is a structured metadata. As a performance optimization this will
+  // only be sent in the first message on the stream.
+  Identifier identifier = 1;
+  // The trace id. this can be used to merge together a streaming trace. Note that the trace_id
+  // is not guaranteed to be spatially or temporally unique.
+  uint64 trace_id = 2;
```
Comment (mattklein123) [inline api/envoy/service/tap/v2alpha/tap.proto]: Is this relevant to a streaming tap where every tap has its own stream?

Comment (yuval-k) [reply api/envoy/service/tap/v2alpha/tap.proto]: yes, if the tap is not buffered, this can help piece together the different pieces of the streamed tap.

File: api/envoy/data/tap/v2alpha/wrapper.proto (hunk 3 above)
Comment (yuval-k) [reply api/envoy/data/tap/v2alpha/wrapper.proto]: yes, it is indeed orthogonal. i will remove this and re-introduce later on.
for reference, the goal is to allow higher-level correlations. for example, allow matching the destination of a tap to a specific pod in a kube cluster.

File: api/envoy/service/tap/v2alpha/common.proto (hunk 4 above)
Comment (yuval-k) [reply api/envoy/service/tap/v2alpha/common.proto]: what about tap_enabled?

Comment (mattklein123) [reply api/envoy/service/tap/v2alpha/common.proto]: Sure +1

**Comment by yuval-k** [comment]
the cat-bot doesn't seem to be able to re-test azure pipelines?

File: api/envoy/config/common/tap/v2alpha/common.proto [hunk 10]
```diff
@@ -14,6 +15,16 @@ option java_package = "io.envoyproxy.envoy.config.common.tap.v2alpha";
 
 // Common configuration for all tap extensions.
 message CommonExtensionConfig {
+
+  // [#not-implemented-hide:]
+  message TapDSConfig {
+    // Configuration for the source of TapDS updates for this Cluster.
+    envoy.api.v2.core.ConfigSource config_source = 1;
```
Comment (mattklein123) [inline api/envoy/config/common/tap/v2alpha/common.proto]: required message validation?

File: api/envoy/service/tap/v2alpha/common.proto [hunk 11]
```diff
@@ -158,6 +171,10 @@ message OutputSink {
 
     // Tap output will be written to a file per tap sink.
     FilePerTapSink file_per_tap = 3;
+
+    // [#not-implemented-hide:]
+    // GrpcService to stream data to. format argument must be PROTO_BINARY.
```
Comment (mattklein123) [inline api/envoy/service/tap/v2alpha/common.proto]:
```suggestion
    // GrpcService to stream data to. The format argument must be PROTO_BINARY.
```

**Comment by mattklein123** [comment]
@yuval-k friendly request to not force push, it makes it much harder to review. Thank you!

**Comment by yuval-k** [comment]
@mattklein123  sorry about the force push - I had to do it to fix DCO, as i used the github interface to accept a suggested fix, which resulted in a commit without a signoff.. not sure how to fix DCO without force push
